### ISSUE TITLE
[Proposal] Add EventTarget.getEventListeners()

### ISSUE THREAD

**Issue by LeaVerou** [whatwg/dom#412]
Originally posted at WICG: https://discourse.wicg.io/t/proposal-eventtarget-prototype-geteventlisteners/2015

**Purpose:** This would provide a way to get all event listeners added to an element via `addEventListener`.

**Signature:** `element.getEventListeners([type])`

**Returns:** Array of objects. Each object contains properties for `type`, `callback`, `options` for the arguments that registered the corresponding event listener. Events registered via inline event handlers are not included.

**Use cases:** This would enable removing events based on arbitrary criteria, instead of requiring a reference to the callback, which causes unnecessary couplings. Typically libraries deal with this by providing their own abstractions for adding events that track the listeners manually. However, this is fragile, as it means listeners not registered via the library cannot be retrieved or removed. Some libraries deal with this by hijacking `addEventListener` to keep track of listeners, but this is very intrusive for a library and it doesn't help with any listeners registered before the library was included. Browsers already keep track of event listeners, so it should be relatively easy to expose them, and is on par with the Extensible Web Manifesto principle of exposing browser "magic" via JS APIs.

**Comment by domenic** [comment]
On balance I think this is probably worth it, but we should note the solid objections to this on a few grounds:

- This opens up the ability for APIs (e.g. custom element APIs) to change their behavior based on the presence or absence of an event listener, which is very bad design. Adding event listeners should be side-effect free.
  - On the other hand, this opens up the ability for libraries to perform unobservable optimizations as side effects of listeners being present, if they are careful.
- This opens up the ability to unregister listeners which you did not register, and thus interfere with other parts of your codebase or other code running on the same page in hard-to-debug ways. The current model has an "object capability" design wherein you can more easily reason about the system. (Note: this is not a security boundary, just a reason-about-code-better boundary, analogous to how you can't resolve or reject a promise unless you are the one that created it.)

I think both of these objections are solid, but as I said, on balance I think they are overcome. The fact that people are hacking around the current design so much, and that this abstraction appears in other platforms (e.g. jQuery's system, or Node.js's EventEmitter) implies to me it's probably better to expose.

It's probable that this subject has been discussed before, maybe pre-GitHub; I wonder if we could find those threads to see how the discussion went then and if anything has changed.

>  Events registered via inline event handlers are not included.

This is weird, but I see the issue; currently the actual listener function is not reified, and it would be unclear what removeEventListener("event", reifiedListener) does (does the content attribute or IDL attribute stick around?).

> Browsers already keep track of event listeners, so it should be relatively easy to expose them, and is on par with the Extensible Web Manifesto principle of exposing browser "magic" via JS APIs.

I think this is a misinterpretation of the situation. Browsers don't have the magic ability to look at all event listeners; just like user code, they only have the ability to dispatch events. The actual event listener list is usually stored as a private variable in the EventTarget backing class, with only equivalents of addEventListener/dispatchEvent exposed to the rest of the codebase. You could always modify a browser to make that private variable public, but that's the same as what we're discussing here for the web. So there's no browser-only magic we're exposing here; this would be a purely new capability, not a piece of bedrock we're unearthing.

**Comment by LeaVerou** [comment]
All good points there, thanks for the thoughtful response @domenic!

On another note, another possible design would be to have the same signature as `addEventListener` with all three arguments (type, callback, options) which would act as filters for which listeners to return. I suspect if the only supported filter is `type`, the rest will be emulated by developers via looping over the returned array.

**Comment by annevk** [comment]
Someone should probably research https://lists.w3.org/Archives/Public/www-dom/ as I recall this coming up quite a few times in the past and being dismissed each time. I don't recall much else unfortunately.

**Comment by LeaVerou** [comment]
This is what I could find:
https://lists.w3.org/Archives/Public/www-dom/2012AprJun/0131.html
https://lists.w3.org/Archives/Public/public-whatwg-archive/2012Jun/0157.html
https://lists.w3.org/Archives/Public/www-dom/2014JulSep/0029.html

At first glance, it appears the pushback is entirely comprised of the theoretical purity arguments that @domenic mentioned would come up. No implementation difficulties or anything of the like.

Tab mentioned an alternative in the last thread: Allowing an option for namespacing and removing events via namespaces, with both strings and symbols allowed. This is a good idea but does not solve all use cases. For example, a major use case is cloning an element with its events. In that case you legitimately want access to listeners you did not bind.

**Comment by annevk** [comment]
I also found https://lists.w3.org/Archives/Public/public-webapi/2008Apr/thread.html#msg66 that isn't very conclusive either way, mostly folks not convinced by the use cases given at the time. (Although allowing iteration over them is somewhat novel I suppose.)

(The identifier grouping idea is #208.)

**Comment by domenic** [comment]
I don't think "theoretical purity" is a fair characterization of those arguments. They are practical arguments from real developers who want to be able to develop components without being afraid that any other piece of code on the page could interfere. The lack of ability to reason about program control flow is an actual issue that affects developers every day, not a theoretical one.

**Comment by domenic** [comment]
For example, given [the code for TacoButton here](https://html.spec.whatwg.org/multipage/scripting.html#custom-elements-autonomous-drawbacks), you'd no longer be able to guarantee that all taco-buttons on the page properly forward enter/space keydowns to click events, or that they stop firing click events when disabled. Being unable to reason about your program in this way is a real burden, and it changes your story as a component author from "include my component and it'll work" to "include my component but make sure all your third-party scripts aren't doing some over-aggressive cleanup action or you'll see strange behavior".

I still lean toward exposing the functionality, and then you'll just have to accept that you can no longer write encapsulated components in that way, but I think it's a real tradeoff between two different types of developer convenience, not one between theoretical purity and developer convenience.

**Comment by ljharb** [comment]
It is an important security property that nobody without a reference to my listener function can remove my listeners.

If a use case is cloning a node with all listeners, perhaps this could be handled with a "copyListeners" API (taking event names as filters, eg) without the need to expose the listener references directly?

What about adding a new "addEventListener" option that opts the listener in to being unprotected? That way the default would remain the same - secure - but users who wanted to waive a bit of security for the convenience of jQuery-like event manipulation could do so?

**Comment by nilssolanki** [comment]
I like the idea of namespacing events; this way, I can remove all my handlers later on without needing to keep their references (which is very helpful when binding functions). As far as I understand, when using a Symbol as a namespace one could create a private namespace which could not be intercepted by third parties (which might be important for developers of third-party code).

E.g.:

```
(() => {
  const namespace = new Symbol('eventNamespace');

  function myFn() {}

  function add() {
    document.addEventListener('click', myFn, { namespace });
  }

  function remove() {
    document.removeEventListener('click', null, { namespace });
  }
});
```

Combined with lhjarbs idea of a `copyListeners` API, this should probably cover most of the use-cases?

**Comment by domenic** [comment]
> It is an important security property that nobody without a reference to my listener function can remove my listeners.

To be clear, it is not a security property, because on the web our threat model is not meant to defend against malicious code running on the same page. It is purely a developer-reasoning-about-code property.

**Comment by LeaVerou** [comment]
Regardless of the characterization of these arguments, it's a false sense of security, since this is possible today by wrapping `addEventListener`. There are libraries that do this. However, I think we both agree that there's a multitude of reasons why this is better done natively than having a library hijack `addEventListener`, right?

**Comment by ljharb** [comment]
It's certainly not an intentional security property, but that does not mean it is not an emergent one.

I can defend against wrapping addEventListener by caching it in first-run code.

I definitely agree it should be done natively - i'm simply suggesting that it be opt in, not by default.

**Comment by LeaVerou** [comment]
> I can defend against wrapping addEventListener by caching it in first-run code.

If you're writing a library, you don't control whether your library is first-run code, so tough luck. If you're not writing a library, then what's the point of defending anything? You control exactly what runs.

**Comment by ljharb** [comment]
I can extract an untainted `addEventListener` from an `iframe`, I can include in the documentation that it needs to be first-run, and often ad network code is ran - last - that I don't control. I don't think it's productive to try to make a claim that making code more robust is a fool's errand; whether it's part of the intentional security model of the web or not.

Is there a reason that an opt-in mechanism (at the `addEventListener` callsite, like how `passive: true` works) wouldn't be a good middle ground, where you could get the functionality you want with zero risk of breaking existing axioms, and zero risk of developers having difficulty reasoning about their code?

**Comment by LeaVerou** [comment]
> Is there a reason that an opt-in mechanism (at the addEventListener callsite, like how passive: true works) wouldn't be a good middle ground, where you could get the functionality you want with zero risk of breaking existing axioms, and zero risk of developers having difficulty reasoning about their code?

One of the most basic HCI (or human psychology in general) principles is that when something is opt-in, few do it. Most of the time because they simply never considered it. Therefore, it renders the feature useless for libraries, and only helps in reducing coupling in one's own code. Opt-out might work, like a `stealth` option for listeners.

**Comment by markandrus** [comment]
@LeaVerou 

> Regardless of the characterization of these arguments, it's a false sense of security, since this is possible today by wrapping `addEventListener`. There are libraries that do this. However, I think we both agree that there's a multitude of reasons why this is better done natively than having a library hijack `addEventListener`, right?

Just my two cents here, but I disagree. As has been pointed out, once you can enumerate event listeners, you can implement something like Node's [`removeAllListeners`](https://nodejs.org/api/events.html#events_emitter_removealllisteners_eventname), which warns

> Note that it is bad practice to remove listeners added elsewhere in the code, particularly when the EventEmitter instance was created by some other component or module (e.g. sockets or file streams).

In practice, I've found this means that, as a library author, if you vend event-emitting objects to your users whose events you also use to drive internal functionality, you need to separate the EventEmitter interface your users will depend on from the one your internal functionality depends on (otherwise, a `removeAllListeners` call could break you, and it's annoying to document around…). I would not like to see the DOM APIs follow suit, and I would prefer encapsulation over convenience in this case.

**Comment by LeaVerou** [comment]
@markandrus The `stealth` option I mentioned above would solve this. I assume UAs would also need this functionality, to protect their own internal listeners. Although I'm not sure why anyone would call `removeAllListeners` without wanting to, you know, remove all listeners.

**Comment by markandrus** [comment]
@LeaVerou a mechanism like `stealth` would be useful 👍  I am happy so long as there is at least an opt-out.

> Although I'm not sure why anyone would call `removeAllListeners` without wanting to, you know, remove all listeners.

There are many different types of developers... I've heard bug reports from some who called this function expecting it to remove only "their" event listeners.

**Comment by rianby64** [comment]
The idea around ```Event.getEventListeners()``` is because you want to clean up the garbage you leaved behind?

**Comment by ibash** [comment]
Having overridden `window.addEventListener` to work around a browser bug, I would welcome this change as a way to make my code more obvious (get event listeners and remove where it makes sense) and more robust (don't have to have my code run first thing to have it work).

**Comment by FremyCompany** [comment]
If you want to copy your event listeners, you can redispatch your events on the old node. That doesn't require destroying the currently existing abstraction.
```
function redirect(event) {
  var ne = event.constructor(event.type, event);
  document.querySelector('input').dispatchEvent(ne);
}
```
https://jsfiddle.net/yrewbnes/

**Comment by annevk** [comment]
@FremyCompany that fiddle doesn't seem to work in Chrome/Firefox/Safari. You'll also lose trustedness of the event that way and target/currentTarget and such.

**Comment by phistuck** [comment]
@annevk - it seems to just be missing a `new` (`new event.constructor(event.type, event);`).

**Comment by dominiccooney** [comment]
From a purely implementation point of view I don't think this would be a terrible hassle to implement. Blink internally uses the same EventListener machinery for some internal use cases, like IndexedDB index population and media controls and things. We'd just need some level of filtering or segregating for them.

I'm not sure about window. Are there any times you have cross origin iframes and different domains get to stick event listeners on the same window through the window proxy or anything like that?

As a recovering? wannabe? web developer I'm mildly anxious about the code that's out there assuming their event listeners, once registered, won't be messed with. On the other hand if you can get a reference to the EventTarget anyway you probably can already cause interesting headaches.

**Comment by rianby64** [comment]
IMHO the only reason to ```getEventListeners()``` is because you want to clean up your own listeners after leaving. But, you can put a little effort in order to track manually your listeners. Also, @domenic said
> ... and it changes your story as a component author from "include my component and it'll work" to "include my component but make sure all your third-party scripts aren't doing some over-aggressive cleanup action or you'll see strange behavior"

For me, this statement has a lot of sense... So, ```getEventListeners()``` doesn't add a significant value to DOM.

**Comment by naugtur** [comment]
I've heard it many times. "If it's in the standard we're _supposed_ to use it because it's the way to go, right?"

I think that's why everyone above is worried about it breaking abstractions.

Calling this method should not seen an obvious choice for a novice developer so that they put in some effort into managing their listeners instead of looping through the list and removing each one when they clean up.

How about making it a "static" method of the DOM node, so that it looks like Object.keys? I'm not sure if hurting discoverability of this feature is really helpful, but noone brought it up before.

Additionally, instead of taking this listing function as a solution for cloning nodes, it'd be cleaner to just have a method for cloning with events. 
Having that available removes the main use case for getting all events. The other major use case is inspection, but that is already available on Dev tools. Might need improvements.

OK. Final thought.
If this goes in, please group the output by some categories that users could set, like "click.myown" to help people use it the right way.

**Comment by LeaVerou** [comment]
Another use case I came across yesterday: I was in a talk about an accessibility tool that automatically adds keyboard shortcuts to existing web apps. If they cannot read whether there are existing key* events on each command, they cannot avoid controls that already have keyboard shortcuts. (although this argues more for a declarative way to specify keyboard shortcuts than for this…)

**Comment by phistuck** [comment]
@LeaVerou - plus, having an event listener does not mean it supports keyboard shortcuts. `accesskey` exists for a declarative way.

**Comment by domenic** [comment]
Small thread hijack: I'd like to draw the attention of people in this thread to the related, but distinct, proposal in #208. Although it doesn't cover all use cases of this thread, it covers some of them, so the audience may be similar. If you work on a library or app that would use the proposal in #208, let us know over in that thread, as implementers are currently trying to judge developer interest.

**Comment by marcysutton** [comment]
Here's another use case I didn't see mentioned here: accessibility auditing tools currently have no API to inspect event listeners, so we can't point out inaccessible UI controls. It's a huge gap in what we can automate for WCAG compliance, and I'd LOVE to see an API for this.

For example, a DIV with a click handler and simple text in it would be a perfect candidate for an accessibility violation. We can look at inline onClick handlers (and we plan to), but not the ones bound in memory. In JavaScript frameworks, many event bindings start their life inline in templates but are removed from the rendered HTML, so we can't sniff those out unless we can evaluate before the page is rendered–not a likely scenario for many accessibility testers.

There are a few potential false positives we've identified, such as a wrapper DIV with a click handler for a larger hit area but legitimately accessible UI controls inside (which is fine), and elements bound through event delegation on a parent element. But having an event detection API would be an excellent starting point.

**Comment by ljharb** [comment]
@marcysutton for your use case, do you need access to the actual listener function, or simply to determine if a handler exists?

**Comment by marcysutton** [comment]
> do you need access to the actual listener function, or simply to determine if a handler exists?

Determining a click handler exists might be enough, but it could help to compare listener functions to filter out false positives.

**Comment by ljharb** [comment]
@marcysutton having any access to the listener function creates [issues](https://github.com/whatwg/dom/issues/412#issuecomment-280697648), but simply detecting its presence would not.

**Comment by dmethvin** [comment]
The data returned by an API like this is going to be tough to interpret. For example, `addEventListener()` can take either a callback function *or* an object that has a `handleEvent()` method.  Also, the same listener can be used in different combination of event types and options including `capture` and `passive` so it might appear several times in the list.

If the goal is to determine whether an element has a listener for a particular event/options combo, the caller (or perhaps the API) would also need to look for listeners on ancestor elements to ensure that the event wasn't handled up the tree. It's possible many listeners won't ever be reached if an earlier listener consistently calls `stopPropagation()` but there's no way to know that from the static list.

As a further confounding variable, it seems like every major web site has click handlers attached at the `document` or `window` level that are used for analytics or ads. I just randomly went to cnn.com in Chrome and selected an element, you can see the results below. 

However, I wonder...since Chrome DevTools has access to this info, can a Chrome plugin get it as well? Would exposing for plugins be a better course of action or at least a first step proof of concept to show that exposing this level of detail in a web platform API is useful in practice?

![capture](https://user-images.githubusercontent.com/157858/33590091-11aecb84-d94b-11e7-9560-dec4e205ea6f.PNG)

**Comment by annevk** [comment]
@ljharb even presence detection is a layering violation and can have bad consequences (though implementations do routinely do this for performance optimizations).

@marcysutton for that use case would an extension not be better? That way you can also audit any third-party resources included in the page.

**Comment by marcysutton** [comment]
@annevk do you mean like a browser extension? Devtools extensions can't be automated, so they won't work in test environments outside of the browser, like tests running in CI and Java integrations. JavaScript loaded into the front-end client to inspect the DOM, outside of an extension, would be the most portable. It's true that extensions allow us to inspect inside of iframes, but we can also do that with Selenium Webdriver. 

Having the API in the browser itself would be pretty critical to exposing rules in a variety of test environments. I wouldn't expect it in JSDOM or PhantomJS since they have more limited DOM APIs, but in headless Chrome for sure.

**Comment by annevk** [comment]
Would that mean that exposing this through Webdriver would address testing? This reminds me of shadow tree encapsulation which is a good practice, but also harms testing as those tools cannot poke into the inner tree. Again, something like Webdriver could expose an expanded API that makes it possible. (If that would work, we could raise it with https://github.com/w3c/webdriver.)

**Comment by marcysutton** [comment]
@annevk we can test within open Shadow DOM, and we've started doing it in [axe-core](https://github.com/dequelabs/axe-core). But users have to be educated to use the open mode if they want to audit for accessibility. So far, it's working okay.

It would be most useful to have an event API in the browser itself so other tools can take advantage without having to reinvent the wheel (extensions and Webdriver, which have browser DOM in common). But I'd be happy to get an API anywhere, frankly, including in extensions and/or Selenium Webdriver. Would Webdriver implementers have any easier of a time dealing working around the issues outlined above, though?

**Comment by annevk** [comment]
Yes, by its nature Webdriver can get access to browser internals we'd rather not expose. And using that would also allow your users to use closed mode for shadow tree, so they don't end up inadvertently depending on the internals of their components.

**Comment by WilcoFiers** [comment]
I think the security concern that was raised, being able to cancel handlers, isn't a particularly strong one. If you want to cancel events and you've got access to the node, all you need to do is clone it and replace the original with the clone. The DOM wasn't built with any sort of boundaries in it. Security boundaries are built around the page, not within parts of it.

I think just in general, testability is improved with this API, not just for axe-core (Which I work on, like Marcy). I think any time you can set something without having the ability to check that it's there is a bad sign for your API. It requires you to track side-effects instead, which leads to brittle tests.

The original argument also has a lot going for it IMO. If just about every major library has to build its own method of tracking events are listened for where, that's a pretty good indication that a feature is missing.

**Comment by annevk** [comment]
There's no security concern, there's an encapsulation concern. It's pretty much identical to that of closed shadow trees, come to think of it.

**Comment by strongholdmedia** [comment]
> For example, given [the code for TacoButton here](https://html.spec.whatwg.org/multipage/scripting.html#custom-elements-autonomous-drawbacks), you'd no longer be able to guarantee that all taco-buttons on the page properly forward enter/space keydowns to click events, or that they stop firing click events when disabled.

One should absolutely not do that at all, for that is bad practice.
"Triggering" click/whatever handlers programmatically is an abysmal idea that should only be practiced when one has absolutely no other possibility (as in "overriding something in a wordpress theme without rewriting/changing it" - and on this level).
_Of course, assistive technologies and a few other such cases may be valid, but nevertheless, the argument stands._

Moreover, one should absolutely not do things like sending some tracking nonsense from <a> click events BEFORE letting the link fire (this alone is the greatest web annoyance ever - what if the library that you are calling did not load? It is absolutely unreasonable from privacy viewpoint to EXPECT it to be there) - this is also an abysmal practice.

This, as I see it, boils down to
* a library developer (or, in reality, anybody) should **not at all** bind event handlers automatically to **any** elements not created/inserted/mounted/whatever by themselves; and
* a library developer (or, in reality, anybody) should **not at all** be expected to have any responsibility for issues arising out of other "developers" unbinding their event handlers from the elements _they_ created.

As far as I am concerned, that be it.

**Comment by LeaVerou** [comment]
@strongholdmedia That's a lot of prescriptiveness with little justification. It's more a list of personal opinions rather than a reasoned argument, and it should be taken as such.

**Comment by FremyCompany** [comment]
@LeaVerou I agree. But @domenic and @annevk's point about encapsulation is a concern I share as well, and I don't think it's a good idea to expose other people's event handlers, they are not meant to be exposed and messed with, that would be a breach of the contract you sign up when you listen to an event. You want the object you observe to call you, but you don't want that object to allow any object with read access to it to be able to call you with untrusted arguments and when no event was fired. If you can get access to the function that listens to an event, you can do just that.

**Comment by FremyCompany** [comment]
I have less of a concern about `hasEventListeners(forEventType:string)` which could be useful for optimization purposes, which also has been proposed in this thread but wouldn't solve the original problem statement and should probably be pursued in a different thread, if desirable.

**Comment by strongholdmedia** [comment]
>  you don't want that object to allow any object with read access to it to be able to call you with untrusted arguments and when no event was fired. If you can get access to the function that listens to 
an event, you can do just that.

I am sorry?
I am fairly sure that at the very moment, you _can_ trigger anyone's event handlers with "untrusted arguments"; and even in the obscure quagmire about makeshift security above, nobody seemed to mention that this would make that possible.

In fact, I just argued against the practice of "triggering" random event handlers (mostly for the sake of perceived simplicity - as in "less work"), e.g. an "onclick" when no actual click happened.

**Comment by strongholdmedia** [comment]
> That's a lot of prescriptiveness with little justification.

You know, I try to get accustomed to recent events in programming, but I am that type who grew up in times when you did not need special reasoning behind points like "you should not routinely falsify a click event when no actual click occurred, just to spare yourself some time".

While I wholeheartedly agree that I did not do much to explain "my" points to complete outsiders, I still insist on they being far from being "personal opinions".

**Comment by FremyCompany** [comment]
@strongholdmedia This is entirely inaccurate. You can dispatch events on an object, but doing so requires sending a specifc Event type which would be typesafe; you cannot single out one specific event handler and call that single handler only. And you certainly cannot call this single event handler with for instance a Proxy instead of an Event, that would get out of the function some things that get passed as argument to functions of the event (or event.target or anything the handler could get from the event object).

Also, please consider the credentials of everyone involved in this thread, and pay a little respect. If you have arguments, feel free to make them, we would love to welcome them, but your comments here are hostile for no good reason, and your only response to "please justify yourself" so far has been "I have good reasons, trust me" and, between you and me, that's a terrible way to answer the question. It would be nice if you could stay objective and on track.

**Comment by strongholdmedia** [comment]
Okay, true that; I did not realize that you were talking about "any single event handler" as opposed to dispatching an event to all handlers.
Based on the "credentials" you mean, it may happen that many of you are not used to the "dark side of event handlers".
I juggle quite frequently with things like "single page event handlers", when somebody, typically with jQuery, implemented some thing inside a handler, then found it easier to just trigger that handler instead of relocating the code inside to an appropriate place, and, ... aaarggh.
But the worst of all is when you cannot open a link with ctrl+click (or.. um.. command, whatever) on a page, because someone put a google tag manager event whatnot on the click event, that redirects instantly after clicking, ctrl or otherwise.
Or when the script is blocked by some privacy-savvy approach, then it returns false, or throws an error, and the site is completely unnavigable (except for contextmenu -> open in new tab).

**Comment by strongholdmedia** [comment]
But, more importantly, I think _this proposal should be implemented_.
If, after all, "user agent vendors" think this be disabled by default, then so it be.
But I think it should at the very least be a cross-platform developer-only feature that is accessible from console.

**Comment by dtapuska** [comment]
Chromium based browsers do support [getEventListeners](https://developers.google.com/web/tools/chrome-devtools/console/utilities#geteventlisteners) via the console.

**Comment by strongholdmedia** [comment]
Yes, that is basically what brought me here.
(Surely enough, I don't use Chromium/Blink, if not for testing.)

**Comment by FremyCompany** [comment]
Thanks for expanding on your opinion; now this forms something that can inform the decision. It's always delicate to call out people when they erroneously assume their opinion is obvious, but you took it to heart, that's nice.

**Comment by strongholdmedia** [comment]
I still don't recognize how recognizing that a link does not work is an opinion, but since it doesn't add much to the core discussion, I guess I leave it to you now.

**Comment by benjamingr** [comment]
Hey, we're discussing an EventTarget in Node.js (thank you to Domenic and Anne for help with the EventTarget/EventEmitter doc!). We are discussing exposing this capability on our EventTarget (by supporting an EventEmitter'ish API).

I wanted to ask for guidance and make sure Node isn't making a huge mistake by exposing it :]

**Comment by strongholdmedia** [comment]
Thank you for the heads-up.
I left a reply there and tried to reflect to some of the points turned up here earlier.

**Comment by RReverser** [comment]
> I have less of a concern about `hasEventListeners(forEventType:string)` which could be useful for optimization purposes, which also has been proposed in this thread but wouldn't solve the original problem statement and should probably be pursued in a different thread, if desirable.

FWIW this would be enough yet important for my usecases as well.

As others noted, it's possible to implement such tracking yourself by tracking counts in `addEventListener`/`removeEventListener`, but it would be quite fragile.

For example, developer needs to carefully handle duplicates in the same way `addEventListener` does. E.g. `addEventListener('foo', foo); addEventListener('foo', foo);` are deduplicated but, say, `addEventListener('foo', foo); e.addEventListener('foo', foo, { capture: true });` are not.

Same applies to removing listeners - when user does `removeEventListener('foo', foo);` you want to only decrease count that matches `addEventListener('foo', foo)` but not `e.addEventListener('foo', foo, { capture: true })`. You also don't want to decrease counts when `removeEventListener` is a no-op because listener was previously removed.

Then, one also needs to correctly handle recently added `signal` option and remove listeners there (but only the "correct" ones as well).

While this is all possible to handle in userland, the amount of complexity of doing it there justifies exposing it as a DOM API instead, as browser already has all the necessary info to return answer for `hasEventListeners` much more efficiently and in a more future-proof manner.

**Comment by strongholdmedia** [comment]
In fact I have got somewhat disgruntled with the sentiment here back then a few years ago, and made a _shim_ a year or so ago, available as `@dorgaren/usable` over at `npm`.

It may not support capture as an options object _properly_, since there is no _simple_ and reliable ES5 way known to me to "hash" an object in a deterministic manner (i.e. without being sensitive to "truthiness", key order, etc.).

But for anyone wishing to do debugging / quick experiments with `hasEventListeners` or even `getEventHandlers` (named differently, due to the syntax not matching the other set of methods), or just have a reasonable event auto-mounting mechanism for attaching event handlers on `DocumentFragment`s and auto-attaching them when the fragment is appended to a document / element, it may be handy whatsoever.

**Comment by WilcoFiers** [comment]
I'm a little sorry to see that 5 years after this was proposed, no progress has been made.

**Comment by annevk** [comment]
Well, I think we made progress in identifying the issues with this proposal early on: https://github.com/whatwg/dom/issues/412#issuecomment-349925053. And thus far encapsulation is not a thing we have been keen to give up on. And I'm not sure there are alternatives that would preserve encapsulation. What kind of progress were you expecting?

**Comment by WebReflection** [comment]
> There's no security concern, there's an encapsulation concern.

Because JS in browsers preserves the string version of any function, dare I say there are security concerns too.

If any app adds a listener that's created on the backend and includes any sort of secret in it that should be theoretically not visible from any other running script (including the possibility to add such listener and then drop right after the script doing that) such function, which closure/scope should never be known out there, might leak some information once evil script getting randomly all listeners from all forms/inputs out there starts looking for specific details within some listener.

One might think this is all theoretical but where I work it's pretty much the daily way to deal with evil scripts (and vice-versa).

Having any listener exposed is, hence, a security concern too, imho.



> It's pretty much identical to that of closed shadow trees, come to think of it.

Exactly, and here web extensions also asked to have higher privileges ... so maybe this method of getting all listeners should be exposed only to web extensions that declare such intent? Yet please consider my previous point.


Last, but not least, the proposal considers only functions as listeners, instead of `{handleEvent(){}}` listeners, which are pretty common out there, and fully rely to the fact nothing without their reference can interfere with their state.

Exposing these kind of listeners will also result in tons of security/encapsulation concerns because that's a way to weakly relate DOM elements to any instance and most libraries using that pattern won't be inherently safe/working properly anymore.


> What kind of progress were you expecting?

Me, personally, a *won't fix*, not because I don't understand the use case, simply because it's been too many years that libraries trust the fact listeners set privately or ASAP will never be reachaed or removed from the node.

The argument about being able to do that anyway in user-land is also weak because I can trap `addEventListener` on top of the scripts and use that directly being sure that nobody can poison that method, hence securing the environment.

This technique is at the core of our injected code in ABP too and it's been rock-solid for years and (up to 100) million devices and we would never want anything to be able to remove a listener we purposely set because the user wants come guarantee.

Thanks for listening.

**Comment by WebReflection** [comment]
I don't think you can exploit a function string representation or its reference in JS so I am not sure I am following. Having malicious scripts able to retrieve listeners is a violation of the current standard where listeners should never leak and securing the env is trivial enough for early/top-most scripts so that nothing can poison prototypes down the road (including function call/apply).

So, maybe security is no the best term, but it's a concern.

> While non-trivial to exploit, programs that run in the same physical process cannot meaningfully keep secrets from each other.

I guess private fields were added for no reason then? I am not sure where this is going 🤔

**Comment by LeaVerou** [comment]
Sorry if this has been mentioned already, but re:security and encapsulation concerns: but developers can already intercept `addEventListener` by doing something like

```js
let addEventListener = EventTarget.prototype.addEventListener;
let removeEventListener = EventTarget.prototype.removeEventListener;

EventTarget.prototype.addEventListener = function(type, callback, options) {
	/* store args and then… */
	addEventListener.call(this, type, callback, options);
};

EventTarget.prototype.removeEventListener = function(type, callback, options) {
	/* remove from stored args and then… */
	removeEventListener.call(this, type, callback, options);
};
```

The problem is just that a) it's flimsy, since listeners may have been attached before the code was run and b) it's suboptimal to be overwriting built-ins like that.

**Comment by WebReflection** [comment]
> developers can already intercept addEventListener by doing something like

yes, it was mentioned, but on the other side, developers can secure nothing will ever interfere with their intent if your code runs earlier (see polyfills)

```js
const {bind: b, call: c} = Function;
const {addEventListener} = EventTarget.prototype;
const addListener = b.call(c, addEventListener);

// any time later on ...
addListener(document.body, 'click', () => { console.log('safe'); });
```

We do this already in explicit circumstances to avoid evil scripts intercepting, removing, or mutating our listeners behaviors, but extensions can guarantee (in some browser) their code executes before anything else.

Again, I think this feature would be a good one for Web Extensions, at least to pave the path, but I also think user land can have a centralized library authors can use to retrieve all listeners, if added/remove through such centralized reference.

That would be also fine, like:

```js
EventTarget.addPublicEventListener(element, listener, options);
```

Now, that would be cool ... changing a very long time expected behavior where many based their libraries around looks like an unnecessary, imho undesired, breaking change I hope won't land as amend of the current methods.

I would be more than OK making it land as new method, but I guess that will also fall into the "*opt-in*" is less effective than "*opt-out*", but then again, if this lands as default behavior, we need to be sure we patch that default behavior and exclude any of our listeners to be exposed in the wild to undesired evil scripts.

**Comment by domenic** [comment]
Any data in the same process can be intercepted with Spectre; there is no security concern. If this is not something you're familiar with then it might be best to read up on that instead of sidetracking this issue. Early running scripts are not protection against that; there is simply no ability to keep secrets within the same process.

The issue is only about encapsulation, not security.

**Comment by WebReflection** [comment]
I’m not sure tracking the issue, I’m talking about real world concern our product, and every similar product, has around leaked listeners.

I wasn’t aware Spectre can be exploited through JavaScript in a browser page and leak every part of the code, but if that’s the only concern around security then fine, let’s talk encapsulation, no reason to be aggressive though.

**Comment by LeaVerou** [comment]
I don't think an opt-in new method would work (if the listener-adding code follow the same "contract" as the listener-reading code they can just use a helper, most use cases for `getEventListeners()` are about doing stuff with listeners you did not attach yourself), but what about an opt-out? I'd wager that _most_ event listeners exist to enable functionality and don't care one iota about encapsulation. For the ones that do, there could be something like `addEventListener(eventName, callback, {closed: true})`

**Comment by benjamingr** [comment]
Let me make sure I am understanding the current situation:

 - There are encapsulation concerns which people agrees with. Adding this API does remove a current (implicit) guarantee that "regular" code cannot access the already registered event listeners.
 - Domenic also raised another very good concern early on - this enables building APIs that rely on an event listener being registered already which is typically not a good design.
 - There seems to be consensus about the use cases Lea raised in the initial post being valid and common. There is also agreement that the current way to accomplish this by altering `EventTarget.prototype` is hacky and imperfect.
 - The idea of having an opt in or opt out for listeners being discoverable by `getEventListeners` has been raised. People seem to agree that having an opt-in reduces the usability of the feature.

What about an alternative "enable encapsulation" options like the ones browsers typically use to enable breaking features like `eval` e.g. a header that restores the "no `getEventListeners`" behavior or an opt-in behavior so most websites can benefit from this but ones who care about this encapsulation cannot.

**Comment by benjamingr** [comment]
Also @WebReflection note in Node.js we already shipped an API that lets you get event listeners to our native `EventTarget` implementation which is also the one we use for our `AbortSignal` and other APIs. It's a static one, to not violate the spec.

**Comment by WebReflection** [comment]
@benjamingr 

> What about an alternative "enable encapsulation" options like the ones browsers typically use to enable breaking features like eval e.g. a header that restores the "no getEventListeners" behavior or an opt-in behavior so most websites can benefit from this but ones who care about this encapsulation cannot?

if this is about special CSP flag, then maybe it makes sense ... but I don't see CSP less cumbersome than opt-in, to be honest, yet it's close to my idea of "*let's bring this to Web Extensions and see how it goes*", so to me that's an option.

> note in Node.js ...

as mentioned, Node.js is a different use case. It doesn't get evil scripts out of the box and it doesn't suffer by design XSS.

> It's a static one, to not violate the spec.

If it's a new method I am not sure how that would violate the specs but again, in node you don't have DOM listeners to deal with, I am not fully sure why Node.js is so relevant in this breaking change for the Web (or the current state).

> Let me make sure I am understanding the current situation:

Yes, all point reflect my understanding too.

**Comment by WebReflection** [comment]
**alternative** how about an explicit API that allows to retrieve all listeners from X time on?

```js
element.addEventListener('click', myOwnBusiness);

const listeners = element.observeListeners();
[...listeners]; // nothing

element.addEventListener('click', publicAffair);
[...listeners]; // [publicAffair]
```

or dare I say, something to implement through a mutation observer when listeners are added/removed? still with a possibility to opt-out from certain listeners (as in filterOut property) ... would this work better?

**Comment by benjamingr** [comment]
CSP can also be opt-out it doesn't have to be opt in (right?).

> alternative how about an explicit API that allows to retrieve all listeners from X time on?

I think anything that doesn't intercept listeners that were there before my code is run is significantly diminished in value.

**Comment by ljharb** [comment]
@benjamingr it is absolutely critical for encapsulation that listeners added before your code runs are not interceptable, for reasons discussed at length upthread.

**Comment by WebReflection** [comment]
I guess I’m echoing @ljharb and yet another reason this feature scares me is custom elements suddenly unable to safely register their listeners to properly work because cowboy.js might suddenly remove all listeners from a node … encapsulation and ownership should be preserved by default as its always been to date. The more I think about this, the more I think opt-in is the best way to move forward, as @ljharb suggested ages ago.

**Comment by strongholdmedia** [comment]
> Me, personally, a _won't fix_, not because I don't understand the use case, simply because it's been too many years that libraries trust the fact listeners set privately or ASAP will never be reachaed or removed from the node.

This is simply not true.
As @LeaVerou noted above, it is well possible since _ages_ to do this.

In fact, I also mentioned my "hijack library" (see `@dorgaren/usable` at npm) for this and similar uses above last year.
(It has a few issues remaining, that are to be fixed soon, the primary one being multiple registration, but as this not being self-promotion, this is not my focus here.)
Loading that at the beginning of a page (either by embedding it into a script on the very same page, _or_ creating a 20-line browser addon that guarantees to load it _before_ any page is loaded).

It makes it facile to retrieve and remove (that is the primary scope) event listeners attached by anyone to anything.

This very ability / behavior, I think, makes most points above by everyone about _expectations_ around some sort of _encapsulation_, as well as _security_ moot.

Any page relying on third party scripts could load something similar (either intentionally or "inadvertently", via hijacking, sandbox vulnerabilities, or any other method) to make this possible for such (or any other) third party scripts (or, as my intention was, for the "JS app" belonging to / being the "page" itself).

Would it be better if this has been implemented natively with some configurable security? Sure, I think.
Is relying on any such stuff for privacy, security, or expectation of actions to be taken / lack thereof unreasonable? Sure, I also think.

**Comment by ljharb** [comment]
@strongholdmedia that's only the case if your hijacking happens before the code meant to be encapsulated; being "first-run" code is the only way to have any sort of guarantees in JS. If event listeners are aded *before* your hijacking code runs, then it must remain impossible for you to observe them. It's totally fine (and impossible to prevent anyways) if later-running code is unable to prevent observation of its event listeners.

**Comment by WebReflection** [comment]
@strongholdmedia what @ljharb said, but I also do that for leaving, I wasn’t here commenting my concerns out of the blue, I use secured code and run it before others daily.

**Comment by bathos** [comment]
wanna echo @ljharb cause his last comment hits on an extremely important general principle of how js "virtualization" works. that other code running first can do this is (kinda confusingly) a key part of the foundation of encapsulation rather than evidence that it's unavailable. but the observability river only runs in one direction. reversing the river on untold amounts of existing code is not okay, but a new method or opt-in flag does not have that problem.

**Comment by WilcoFiers** [comment]
Instead of returning a function that might potentially include secrets in its string, could some representation of that function be returned instead?

**Comment by benjamingr** [comment]
Just another point I thought about: Even with `getEventListeners()` it is still possible for code interested in doing so to hide event handlers and prevent removing them from consumers by overriding `EventTarget.prototype.getEventListeners()` itself before any external code can get a reference.

```js
const origGetEventListeners = EventTarget.prototype.getEventListeners;
Object.defineProperty(EventTarget.prototype, 'getEventListeners', {
  value: function myGetEventListeners(eventType) {
    return origGetEventListeners.apply(this, arguments).filter(handler => myCustomFilterLogic(handler));
  }
});
```

So encapsulation is still possible just not the default with Lea's proposal.

**Comment by ljharb** [comment]
@benjamingr youd probably also have to prevent/hijack creation of new iframes and opened windows, all of which would have a fresh copy of getAllEventListeners. The scope of the problem changes pretty drastically if this method is available.

**Comment by benjamingr** [comment]
> youd probably also have to prevent/hijack creation of new iframes and opened windows, all of which would have a fresh copy of getAllEventListeners.

It's probably(?) fine to make `getEventListeners` throw an error if it was used on a different realm which would resolve that.

**Comment by ljharb** [comment]
Do any other web methods do that? That seems like it’d be a pretty big inconsistency to introduce.

**Comment by bathos** [comment]
> So encapsulation is still possible just not the default with Lea's proposal.

It would be ... for code that’s yet to be written.

**Comment by WebReflection** [comment]
@benjamingr I think every “just pollute native prototype” argument could be used, and has been used, already to indeed solve the original issue. For the same reason that doesn’t work one way, it doesn’t any other way and it will also defeat the intended goal of this proposal. Now imagine every library trusting secured events doing that trick …

**Comment by dmethvin** [comment]
Assuming there is a link in the document, what is the value of `x` in this example?
```
document.body.addEventListener("click", (event) => console.log("I am listening"));
var x = document.getElementsByTagName("a")[0].getEventListeners("click");
```

**Comment by rlconst** [comment]
@dmethvin `[ƒ]`

**Comment by subhanahmed047** [comment]
> Sorry if this has been mentioned already, but re:security and encapsulation concerns: but developers can already intercept `addEventListener` by doing something like
> 
> ```js
> let addEventListener = EventTarget.prototype.addEventListener;
> let removeEventListener = EventTarget.prototype.removeEventListener;
> 
> EventTarget.prototype.addEventListener = function(type, callback, options) {
> 	/* store args and then… */
> 	addEventListener.call(this, type, callback, options);
> };
> 
> EventTarget.prototype.removeEventListener = function(type, callback, options) {
> 	/* remove from stored args and then… */
> 	removeEventListener.call(this, type, callback, options);
> };
> ```
> 
> The problem is just that a) it's flimsy, since listeners may have been attached before the code was run and b) it's suboptimal to be overwriting built-ins like that.

I don't think the `/* remove from stored args and then… */` part works when you are using the abort controller to remove an event listener. i.e.
```
const controller = new AbortController();
document.addEventListener('clicl', () => {}, { signal: controller.signal });

controller.abort();
```

**Comment by annevk** [comment]
I'm going to close this as I don't think the original idea of exposing the listeners as-is is something we ever want to do as breaking people's existing encapsulation setups would not be okay. As per https://dom.spec.whatwg.org/#observing-event-listeners observing event listeners is an anti-pattern.

If anyone is keen to reopen this topic in some way I recommend keeping https://whatwg.org/faq#adding-new-features in mind.
